### PR TITLE
Move nodegroup balancing to processor, add GKE-specific implementation

### PR DESCRIPTION
**Pull request by MaciekPytel**
This allows customizing the balancing logic for different use-cases. In particular this PR implements GKE specific version (only enabled if provider is gke) that considers node groups with the same gke-nodepool as similar.

### CODE REVIEW

**Comment by MaciekPytel** [comment]
@losipiuk Refactored based on our conversation, PTAL.

File: cluster-autoscaler/processors/nodegroupset/balancing_processor.go [hunk 1]
```diff
@@ -17,30 +17,54 @@ limitations under the License.
...
+		if ngId == nodeGroupId {
+			continue
+		}
+		ngNodeInfo, found := nodeInfosForGroups[ngId]
+		if !found {
+			glog.Warningf("Failed to find nodeInfo for group %v", ngId)
+			continue
+		}
+		comparator := b.Comparator
+		if comparator == nil {
```
Comment (losipiuk) [inline cluster-autoscaler/processors/nodegroupset/balancing_processor.go]: I would rather not have this logic here and treat Comparator as obligatory paremetrization.

File: cluster-autoscaler/processors/nodegroupset/balancing_processor.go [hunk 2]
```diff
@@ -17,30 +17,54 @@ limitations under the License.
...
+		ngId := ng.Id()
+		if ngId == nodeGroupId {
+			continue
+		}
+		ngNodeInfo, found := nodeInfosForGroups[ngId]
+		if !found {
+			glog.Warningf("Failed to find nodeInfo for group %v", ngId)
+			continue
+		}
+		comparator := b.Comparator
```
Comment (losipiuk) [inline cluster-autoscaler/processors/nodegroupset/balancing_processor.go]: why copy to variable?

File: cluster-autoscaler/processors/nodegroupset/compare_nodegroups_test.go [hunk 3]
```diff
@@ -29,22 +29,22 @@ import (
...
 	ni2.SetNode(n2)
-	assert.Equal(t, shouldEqual, IsNodeInfoSimilar(ni1, ni2))
+	assert.Equal(t, shouldEqual, comparator(ni1, ni2))
 }
 
 func TestIdenticalNodesSimilar(t *testing.T) {
 	n1 := BuildTestNode("node1", 1000, 2000)
 	n2 := BuildTestNode("node2", 1000, 2000)
-	checkNodesSimilar(t, n1, n2, true)
+	checkNodesSimilar(t, n1, n2, IsNodeInfoSimilar, true)
```
Comment (losipiuk) [inline cluster-autoscaler/processors/nodegroupset/compare_nodegroups_test.go]: what is the reason for passing same comparator for every test?
Drop argument?

Comment (losipiuk) [reply cluster-autoscaler/processors/nodegroupset/compare_nodegroups_test.go]: Ok I see the other test.

File: cluster-autoscaler/processors/nodegroupset/gke_compare_nodegroups.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// IsGkeNodeInfoSimilar compares if two nodes should be considered part of the
+// same NodeGroupSet. This is true if they either belong to the same GKE nodepool
+// or match usual conditions checked by IsNodeInfoSimilar.
+func IsGkeNodeInfoSimilar(n1, n2 *schedulercache.NodeInfo) bool {
+	if nodesFromSameGkeNodePool(n1, n2) {
+		return true
+	}
+	return IsNodeInfoSimilar(n1, n2)
```
Comment (losipiuk) [inline cluster-autoscaler/processors/nodegroupset/gke_compare_nodegroups.go]: Or maybe return false? WDYT? Too much a change of semantics?

File: cluster-autoscaler/processors/nodegroupset/balancing_processor.go (hunk 2 above)
Comment (MaciekPytel) [reply cluster-autoscaler/processors/nodegroupset/balancing_processor.go]: I think it's a standard coding pattern for applying defaults :p

**Comment by losipiuk** [comment]
@MaciekPytel unhold if you do are not willing to do any changes. Everything is optional from my side.

**Pull request merged**